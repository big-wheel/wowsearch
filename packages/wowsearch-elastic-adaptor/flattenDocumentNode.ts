/**
 * @file flattenDocumentNode
 * @author Cuttle Cong
 * @date 2019/9/23
 *
 */
import DocumentNode from 'wowsearch-parse/types/DocumentNode'
import * as visit from '@moyuyc/visit-tree'
import ContentNode from 'wowsearch-parse/types/ContentNode'
import LvlNode from 'wowsearch-parse/types/LvlNode'
import { parseLvlTypeLevel } from 'wowsearch-parse'
import * as template from 'lodash.template'

export type FlattenedNode = {
  anchor: string
  url: string
  level: number
  type: 'text' | 'lvl' | any
  content: string
  parents: string[]
  // for filter
  tags: string[]
}

export default (documentNode: DocumentNode, {
  url_tpl = '${url}#${anchor}'
} = {}) => {
  let nodes = [] as FlattenedNode[]
  let parentsMap = new WeakMap<ContentNode, string[]>()

  let urlGetter = template(url_tpl)
  visit(
    documentNode,
    (node: ContentNode, ctx) => {
      const anchor = (node as LvlNode).anchor
      const flattenedNode = {
        anchor,
        url: urlGetter({url: documentNode.href, anchor: anchor || '', ...node}),
        level: (node as LvlNode).level,
        type: node.type,
        content: node.value,
        parents: parentsMap.get(ctx.parent) || [],
        tags: []
      } as FlattenedNode
      const parents = flattenedNode.parents.slice()
      if (node.type === 'document') {
        const global = (node as DocumentNode).global || new Map()
        for (const [key, value] of global.entries()) {
          const level = parseLvlTypeLevel(key)
          if (level !== null) {
            parents[level] = value
          }
          nodes.push({
            ...flattenedNode,
            content: value,
            level
          })
        }
      } else {
        nodes.push(flattenedNode)
      }

      parentsMap.set(node, parents.concat(node.value).filter(Boolean))
      // nodes.push({
      //   ...flattenedNode,
      //   content: value
      // })
      //
      // parentsMap.set(node, parents)
    },
    {
      path: 'children'
    }
  )

  return nodes
}
