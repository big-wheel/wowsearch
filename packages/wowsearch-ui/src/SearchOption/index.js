/**
 * @file index
 * @author imcuttle <moyuyc95@gmail.com>
 * @date 2019/9/26
 *
 */
import React from 'react'
import p from 'prefix-classname'

const c = p('')
const cn = p('wowsearch-search-option-')

export default function SearchOption({className, openInNew, title, crumbs = [], url, content}) {
  return (
    <a onClick={evt => {
      if (!evt.metaKey) {
        // 不走 a 默认的跳转逻辑
        evt.preventDefault()
      } else {
        evt.stopPropagation()
      }
    }} target={openInNew ? '_blank' : '_self'} href={url} className={c(cn('wrapper'), className)}>
      <div className={cn('left')} dangerouslySetInnerHTML={{__html: title}} />
      <div className={cn('right')}>
        <div className={cn('header')}>
          {crumbs.map((crumb, i) => {
            const preCrumb = i > 0 ? ' > ' : ''
            return <span key={i} className={cn('crumb')} dangerouslySetInnerHTML={{__html: preCrumb + crumb}} />
          })}
        </div>
        <div className={cn('body')} dangerouslySetInnerHTML={{__html: content}} />
      </div>
    </a>
  )
}
