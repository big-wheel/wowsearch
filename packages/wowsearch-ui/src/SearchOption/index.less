@prefix: ~'wowsearch-search-option-';

.@{prefix} {
  &wrapper {
    display: flex;
  }

  &left {
    flex: 1;
  }

  &right {
    flex: 3;
    display: flex;
    flex-direction: column;
  }
}
