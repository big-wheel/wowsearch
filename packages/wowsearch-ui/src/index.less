//@import "animation.less";
@select-prefix: ~'wowsearch-ui-search';
@select-highlight-color: #4c84ff;
//@select-highlight-bg-color: alpha(lighten(@select-highlight-color, .7), .6);
@select-highlight-bg-color: rgba(143, 187, 237, 0.1);

@search-prefix: ~'wowsearch-search-option';

.word-break() {
  // ff
  word-break: break-all;
  // chrome
  word-break: break-word;

  overflow-wrap: break-word;
}

.@{select-prefix} {
  &-highlight {
    display: inline;
    color: @select-highlight-color;
    background-color: @select-highlight-bg-color;
  }
  &-category-header {
    padding: 8px 0;
    border-bottom: 1px solid #d9d9d9;
  }
}

.search-input-without-border() {
  .@{select-prefix}-selection-search-input {
    border: none;
    outline: none;
    //background: rgba(255, 0, 0, 0.2);
    width: 100%;
  }
}

.@{select-prefix} {
  display: inline-block;
  font-size: 12px;
  width: 100%;
  position: relative;

  &-selection-search-input {
    padding: 0;
    background-color: transparent;
  }

  &-disabled {
    &,
    & input {
      cursor: not-allowed;
    }

    .@{select-prefix}-selector {
      opacity: 0.3;
    }
  }

  &-show-arrow&-loading {
    .@{select-prefix}-arrow {
      &-icon::after {
        box-sizing: border-box;
        width: 12px;
        height: 12px;
        border-radius: 100%;
        border: 2px solid #999;
        border-top-color: transparent;
        border-bottom-color: transparent;
        transform: none;
        margin-top: 4px;

        animation: rcSelectLoadingIcon 0.5s infinite;
      }
    }
  }

  // ============== Selector ===============
  .@{select-prefix}-selection-placeholder {
    opacity: 0.4;
    padding-left: 12px;
  }

  // --------------- Single ----------------
  &-single {
    .@{select-prefix}-selector {
      //display: flex;
      position: relative;

      .@{select-prefix}-selection-search {
        width: 100%;

        &-input {
          width: 100%;
        }
      }

      .@{select-prefix}-selection-item,
      .@{select-prefix}-selection-placeholder {
        position: absolute;
        top: 1px;
        left: 3px;
        pointer-events: none;
      }
    }

    // Not customize
    &:not(.@{select-prefix}-customize-input) {
      .@{select-prefix}-selector {

        height: 25px;
        line-height: 25px;

        padding: 0 12px;
        border: 1px solid #d9d9d9;

        border-radius: 4px;

        .search-input-without-border();
      }
    }
  }

  // -------------- Multiple ---------------
  &-multiple .@{select-prefix}-selector {
    //display: flex;
    //flex-wrap: wrap;
    padding: 1px;
    border: 1px solid #000;

    .@{select-prefix}-selection-item {
      flex: none;
      background: #bbb;
      border-radius: 4px;
      margin-right: 2px;
      padding: 0 8px;

      &-disabled {
        cursor: not-allowed;
        opacity: 0.5;
      }
    }

    .@{select-prefix}-selection-search {
      position: relative;

      &-input,
      &-mirror {
        padding: 1px;
        font-family: system-ui;
      }

      &-mirror {
        position: absolute;
        z-index: 999;
        white-space: nowrap;
        left: 0;
        top: 0;
        visibility: hidden;
      }
    }

    .search-input-without-border();
  }

  // ================ Icons ================
  &-allow-clear {
    &.@{select-prefix}-multiple .@{select-prefix}-selector {
      padding-right: 20px;
    }

    .@{select-prefix}-clear {
      position: absolute;
      right: 20px;
      top: 0;
    }
  }

  &-show-arrow {
    &.@{select-prefix}-multiple .@{select-prefix}-selector {
      padding-right: 20px;
    }

    .@{select-prefix}-arrow {
      pointer-events: none;
      position: absolute;
      right: 5px;
      top: 4px;

      &-icon::after {
        content: '';
        border: 5px solid transparent;
        width: 0;
        height: 0;
        display: inline-block;
        border-top-color: #999;
        transform: translateY(5px);
      }
    }
  }

  // =============== Focused ===============
  &-focused {
    .@{select-prefix}-selector {
      border-color: @select-highlight-color;
    }
  }

  // ============== Dropdown ===============
  &-dropdown {
    min-height: 100px;
    position: absolute;
    background: #fff;


    min-width: 500px;
    //max-width: 600px;

    border: 1px solid #d9d9d9;
    border-radius: 4px;
    box-shadow: 0 1px 0 0 rgba(0, 0, 0, 0.2), 0 2px 3px 0 rgba(0, 0, 0, 0.1);

    &-hidden {
      display: none;
    }
  }

  // =============== Option ================
  &-item {
    font-size: 16px;
    line-height: 1.5;
    padding: 0 16px;

    // >>> Group
    &-group {
      color: #999;
      font-weight: bold;
      font-size: 80%;
    }

    // >>> Option
    &-option {
      position: relative;

      &-grouped {
        padding-left: 24px;
      }

      // search option detail
      .@{search-prefix} {
        &-wrapper {
          color: #666;
          text-decoration: none;
          margin-top: 10px;
        }

        &-left {
          .word-break();
          text-align: right;
          padding: 6px 12px;
        }

        &-right {
          .word-break();
          border-left: 1px solid #ddd;
          padding: 6px 12px;
          color: #314659;
          font-weight: bold;
        }

        &-body {
          font-size: 14px;
          color: #63676c;
          font-weight: normal;
        }
      }

      .@{select-prefix}-item-option-state {
        position: absolute;
        right: 0;
        top: 4px;
        pointer-events: none;
      }

      // ------- Active -------
      &-active {
        .@{search-prefix}-right {
          background: #F5F9FD;
        }

      }

      // ------ Disabled ------
      &-disabled {
        color: #999;
      }
    }

    // >>> Empty
    &-empty {
      text-align: center;
      color: #999;
      padding-top: 10px;
      padding-bottom: 10px;
    }
  }
}

.@{select-prefix}-selection__choice-zoom {
  transition: all 0.3s;
}

.@{select-prefix}-selection__choice-zoom-appear {
  opacity: 0;
  transform: scale(0.5);

  &&-active {
    opacity: 1;
    transform: scale(1);
  }
}
.@{select-prefix}-selection__choice-zoom-leave {
  opacity: 1;
  transform: scale(1);

  &&-active {
    opacity: 0;
    transform: scale(0.5);
  }
}

.effect() {
  animation-duration: 0.3s;
  animation-fill-mode: both;
  transform-origin: 0 0;
}

.@{select-prefix}-dropdown {
  &-slide-up-enter,
  &-slide-up-appear {
    .effect();
    opacity: 0;
    animation-timing-function: cubic-bezier(0.08, 0.82, 0.17, 1);
    animation-play-state: paused;
  }

  &-slide-up-leave {
    .effect();
    opacity: 1;
    animation-timing-function: cubic-bezier(0.6, 0.04, 0.98, 0.34);
    animation-play-state: paused;
  }

  &-slide-up-enter&-slide-up-enter-active&-placement-bottomLeft,
  &-slide-up-appear&-slide-up-appear-active&-placement-bottomLeft {
    animation-name: rcSelectDropdownSlideUpIn;
    animation-play-state: running;
  }

  &-slide-up-leave&-slide-up-leave-active&-placement-bottomLeft {
    animation-name: rcSelectDropdownSlideUpOut;
    animation-play-state: running;
  }

  &-slide-up-enter&-slide-up-enter-active&-placement-topLeft,
  &-slide-up-appear&-slide-up-appear-active&-placement-topLeft {
    animation-name: rcSelectDropdownSlideDownIn;
    animation-play-state: running;
  }

  &-slide-up-leave&-slide-up-leave-active&-placement-topLeft {
    animation-name: rcSelectDropdownSlideDownOut;
    animation-play-state: running;
  }
}

@keyframes rcSelectDropdownSlideUpIn {
  0% {
    opacity: 0;
    transform-origin: 0% 0%;
    transform: scaleY(0);
  }
  100% {
    opacity: 1;
    transform-origin: 0% 0%;
    transform: scaleY(1);
  }
}
@keyframes rcSelectDropdownSlideUpOut {
  0% {
    opacity: 1;
    transform-origin: 0% 0%;
    transform: scaleY(1);
  }
  100% {
    opacity: 0;
    transform-origin: 0% 0%;
    transform: scaleY(0);
  }
}

@keyframes rcSelectLoadingIcon {
  0% {
    transform: rotate(0);
  }
  100% {
    transform: rotate(360deg);
  }
}

