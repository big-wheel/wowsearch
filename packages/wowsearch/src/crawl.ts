/**
 * @file crawl
 * @author Cuttle Cong
 * @date 2018/6/10
 * @description
 */
import { CrawlerConfig } from 'wowsearch-parse/types/Config'
import * as fetch from 'isomorphic-fetch'
import { JSDOM } from 'jsdom'
import * as each from 'lodash.foreach'

import makeCheck from './makeCheckUrl'
import * as u from 'url'
import parseElementTree from 'wowsearch-parse'
import selectVal, { selectAll } from 'wowsearch-parse/selectVal'
import DocumentNode from 'wowsearch-parse/types/DocumentNode'

const debug = require('debug')('wowsearch:crawl')

const protocolPorts = new Map<any, number>([
  ['https:', 443],
  ['http:', 80],
  [null, null]
])

export function isSameOrigin(valueUrl: string, fromUrl?: string) {
  valueUrl = valueUrl.trim()
  let { pathname, protocol, hostname, port } = u.parse(valueUrl)
  if (valueUrl.startsWith('//') || protocol == null) {
    return true
  }
  if (!fromUrl) {
    return false
  }
  const fo = u.parse(fromUrl)
  fo.port = fo.port || <any>protocolPorts.get(fo.protocol)
  port = port || <any>protocolPorts.get(protocol)
  return (
    fo.protocol === protocol && fo.hostname === hostname && fo.port === port
  )
}

function getCrawlingUrls(
  document,
  { fromUrl = '', config }: { fromUrl?: string; config?: CrawlerConfig } = {}
) {
  const { start_urls, stop_urls, force_crawling_urls } = config

  const check = makeCheck({ start_urls: start_urls.concat('**'), stop_urls })
  const smartCrawlingUrls = []
  ;[].slice.apply(document.querySelectorAll('a')).map(function(node) {
    let href = node.getAttribute('href')
    if (href && isSameOrigin(href, fromUrl)) {
      href = u.resolve(fromUrl, href)
      let o = u.parse(href)
      // delete o.query
      delete o.search
      delete o.hash
      href = u.format(o)

      debug('checked href: %s, isSameOrigin', href)
      if (force_crawling_urls || check(href)) {
        fromUrl !== href &&
          smartCrawlingUrls.indexOf(href) < 0 &&
          smartCrawlingUrls.push(href)
      }
    } else {
      debug('checked href: %s, not matched', href)
    }
  })

  debug('smartCrawlingUrls: %o, from url: %s', smartCrawlingUrls, fromUrl)
  return smartCrawlingUrls
}

type CrawlResult = {
  documentNode?: DocumentNode
  smartCrawlingUrls?: string[]
}

export async function crawl(
  text: string,
  config: CrawlerConfig,
  fromUrl?
): Promise<CrawlResult> {
  const { document } = new JSDOM(text, { url: fromUrl }).window
  const {
    start_urls,
    stop_urls,
    strip_chars,
    selectors,
    selectors_exclude,
    smart_crawling,
    url_tpl,
    source_adaptor,
    force_crawling_urls
  } = config

  let smartCrawlingUrls = []
  if (smart_crawling) {
    smartCrawlingUrls = getCrawlingUrls(document, { fromUrl, config })
  }

  // TODO
  // Do filter about start_urls and stop_urls.
  if (selectors_exclude) {
    each(selectors_exclude, (val, key) => {
      const nodes = selectAll(document, val)
      nodes.forEach(node => node.remove())
    })
  }

  const documentNode = parseElementTree(document, selectors)
  documentNode.href = fromUrl

  return {
    documentNode,
    smartCrawlingUrls
  }
}

export async function crawlByUrl(
  url: string,
  config: CrawlerConfig
): Promise<CrawlResult> {
  let html
  if (config.js_render) {
    const puppeteer = require('puppeteer')
    const browser = await puppeteer.launch({
      timeout: 100000,
      args: ['--no-sandbox', '--disable-setuid-sandbox']
    })
    const page = await browser.newPage()
    await page.goto(url, { waitUntil: 'networkidle2', timeout: 100000 })
    config.js_waitfor && (await page.waitFor(config.js_waitfor))
    html = await page.evaluate(() => {
      return document.documentElement.outerHTML
    })
    await browser.close()
  } else {
    try {
      let res = await fetch(url)
      html = await res.text()
    } catch (e) {
      console.error(`URL: ${url}`, String(e))
      return null
    }
  }
  debug('html: %s', html)

  return crawl(html, config, url)
}

export async function pushDocumentNode(
  documentNode: DocumentNode,
  config: CrawlerConfig
) {
  const { url_tpl, source_adaptor } = config

  // Push
  const pushAdaptor = require(source_adaptor.name)(source_adaptor.options)
  const result = await pushAdaptor(documentNode, config)
  return result !== false
}

export async function push(text: string, config: CrawlerConfig, fromUrl?) {
  const { documentNode } = await crawl(text, config, fromUrl)
  return pushDocumentNode(documentNode, config)
}

export async function pushByUrl(url: string, config: CrawlerConfig) {
  const { documentNode } = await crawlByUrl(url, config)
  return pushDocumentNode(documentNode, config)
}
