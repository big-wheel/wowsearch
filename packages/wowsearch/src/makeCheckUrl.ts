import { isRule, Rule } from 'wowsearch-parse/dist/match'
import match from 'wowsearch-parse/dist/match'

/**
 * @file makeCheckUrl
 * @author Cuttle Cong
 * @date 2018/6/10
 * @description
 */

export function makeIsInStartUrls(start_urls) {
  return function isInStartUrls(url) {
    return start_urls.some(rule => {
      if (isRule(rule)) {
        return match(<Rule>rule, url)
      }
      return match((<any>rule).url, url)
    })
  }
}

export default function check(start_urls = [], stop_urls = []) {
  const isInStartUrls = makeIsInStartUrls(start_urls)
  const isInStopUrls = makeIsInStopUrls(stop_urls)
  return function check(url) {
    return isInStartUrls(url) && !isInStopUrls(url)
  }
}

export function makeIsInStopUrls(stop_urls) {
  return function isInStopUrls(url) {
    return stop_urls.some(rule => {
      return match(<Rule>rule, url)
    })
  }
}
