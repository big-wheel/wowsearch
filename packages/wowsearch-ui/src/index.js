/**
 * @file index
 * @author imcuttle <moyuyc95@gmail.com>
 * @date 2019/9/26
 *
 */
import React from 'react'
import Select, { Option, OptGroup } from 'rc-select'
import './index.less'
import { selectPrefix } from '!get-less-vars/loader!./index.less'
import * as PropTypes from 'prop-types'
import {robust} from 'memoize-fn'
import SearchOption from './SearchOption'

SearchUI.propTypes = {
  fetcher: PropTypes.func.isRequired
}

export default function SearchUI({ fetcher }) {
  const mFetcher = React.useMemo(() => {
    return robust(fetcher, {})
  }, [fetcher])

  const [dataSource, setDataSource] = React.useState([])
  const fetchData = async value => {
    // todo error handler
    const list = await mFetcher(value)
    setDataSource(list)
  }

  const onChange = value => {

  }
  const options = dataSource.map(d => {
    return (
      <Option key={d.id}>
        <SearchOption {...d}/>
      </Option>
    )
  })

  return (
    <Select
      labelInValue
      prefixCls={selectPrefix}
      onSearch={fetchData}
      showSearch
      value={null}
      optionLabelProp="children"
      placeholder="placeholder"
      defaultActiveFirstOption
      style={{ width: 500 }}
      onChange={onChange}
      filterOption={false}
    >
      {options}
    </Select>
  )
}
