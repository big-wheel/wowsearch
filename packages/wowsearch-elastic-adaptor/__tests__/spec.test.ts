/**
 * @file spec
 * @author imcuttle
 * @description
 */
import { crawl } from 'wowsearch/src/crawl'
import { readFileSync } from 'fs'
import { makeFixture } from './help'
import flattenDocumentNode from '../flattenDocumentNode'


describe('flattenDocumentNode', function() {
  it('should flattenDocumentNode spec', async function() {
    const { documentNode } = await crawl(
      readFileSync(makeFixture('text.html')).toString(),
      {
        // strip_chars: 'r',
        selectors: {
          lvl0: '#readme .package-name-redundant',
          lvl1: '#readme h1:not(.package-name-redundant)',
          lvl2: {
            selector: '//*[@id="readme"]//h2',
            type: 'xpath'
          },
          lvl3: {
            selector: '#readme h3'
          },
          lvl4: { selector: '#readme h4', default_value: 'abc' },
          text: '#readme > p, #readme li, #readme code, #readme pre'
        },
        selectors_exclude: ['#readme .deep-link']
      }
    )

    expect(flattenDocumentNode(documentNode)).toMatchSnapshot()
  })
})
