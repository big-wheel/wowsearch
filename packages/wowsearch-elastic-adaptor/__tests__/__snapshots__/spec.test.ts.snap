// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`flattenDocumentNode should flattenDocumentNode spec 1`] = `
Array [
  Object {
    "anchor": null,
    "content": "robots.js",
    "level": 0,
    "parents": Array [],
    "tags": Array [],
    "type": "lvl",
    "url": "#",
  },
  Object {
    "anchor": undefined,
    "content": "robots.js — is parser for robots.txt files for node.js.",
    "level": undefined,
    "parents": Array [
      "robots.js",
    ],
    "tags": Array [],
    "type": "text",
    "url": "#",
  },
  Object {
    "anchor": null,
    "content": "Installation",
    "level": 2,
    "parents": Array [
      "robots.js",
    ],
    "tags": Array [],
    "type": "lvl",
    "url": "#",
  },
  Object {
    "anchor": undefined,
    "content": "It's recommended to install via npm:",
    "level": undefined,
    "parents": Array [
      "robots.js",
      "Installation",
    ],
    "tags": Array [],
    "type": "text",
    "url": "#",
  },
  Object {
    "anchor": undefined,
    "content": "$ npm install -g robots",
    "level": undefined,
    "parents": Array [
      "robots.js",
      "Installation",
    ],
    "tags": Array [],
    "type": "text",
    "url": "#",
  },
  Object {
    "anchor": null,
    "content": "Usage",
    "level": 2,
    "parents": Array [
      "robots.js",
    ],
    "tags": Array [],
    "type": "lvl",
    "url": "#",
  },
  Object {
    "anchor": undefined,
    "content": "Here's an example of using robots.js:",
    "level": undefined,
    "parents": Array [
      "robots.js",
      "Usage",
    ],
    "tags": Array [],
    "type": "text",
    "url": "#",
  },
  Object {
    "anchor": undefined,
    "content": "var robots = require('robots')  , parser = new robots.RobotsParser(); parser.setUrl('http://nodeguide.ru/robots.txt', function(parser, success) {  if(success) {    parser.canFetch('*', '/doc/dailyjs-nodepad/', function (access) {      if (access) {        // parse url      }    });  }});",
    "level": undefined,
    "parents": Array [
      "robots.js",
      "Usage",
    ],
    "tags": Array [],
    "type": "text",
    "url": "#",
  },
  Object {
    "anchor": undefined,
    "content": "Default crawler user-agent is:",
    "level": undefined,
    "parents": Array [
      "robots.js",
      "Usage",
    ],
    "tags": Array [],
    "type": "text",
    "url": "#",
  },
  Object {
    "anchor": undefined,
    "content": "Mozilla/5.0 (X11; Linux i686; rv:5.0) Gecko/20100101 Firefox/5.0
",
    "level": undefined,
    "parents": Array [
      "robots.js",
      "Usage",
    ],
    "tags": Array [],
    "type": "text",
    "url": "#",
  },
  Object {
    "anchor": undefined,
    "content": "Here's an example of using another user-agent and more detailed callback:",
    "level": undefined,
    "parents": Array [
      "robots.js",
      "Usage",
    ],
    "tags": Array [],
    "type": "text",
    "url": "#",
  },
  Object {
    "anchor": undefined,
    "content": "var robots = require('robots')  , parser = new robots.RobotsParser(                'http://nodeguide.ru/robots.txt',                'Mozilla/5.0 (compatible; RobotTxtBot/1.0)',                after_parse            );            function after_parse(parser, success) {  if(success) {    parser.canFetch('*', '/doc/dailyjs-nodepad/', function (access, url, reason) {      if (access) {        console.log(' url: '+url+', access: '+access);        // parse url ...      }    });  }};",
    "level": undefined,
    "parents": Array [
      "robots.js",
      "Usage",
    ],
    "tags": Array [],
    "type": "text",
    "url": "#",
  },
  Object {
    "anchor": undefined,
    "content": "Here's an example of getting list of sitemaps:",
    "level": undefined,
    "parents": Array [
      "robots.js",
      "Usage",
    ],
    "tags": Array [],
    "type": "text",
    "url": "#",
  },
  Object {
    "anchor": undefined,
    "content": "var robots = require('robots')  , parser = new robots.RobotsParser(); parser.setUrl('http://nodeguide.ru/robots.txt', function(parser, success) {  if(success) {    parser.getSitemaps(function(sitemaps) {      // sitemaps — array    });  }});",
    "level": undefined,
    "parents": Array [
      "robots.js",
      "Usage",
    ],
    "tags": Array [],
    "type": "text",
    "url": "#",
  },
  Object {
    "anchor": undefined,
    "content": "Here's an example of getCrawlDelay usage:",
    "level": undefined,
    "parents": Array [
      "robots.js",
      "Usage",
    ],
    "tags": Array [],
    "type": "text",
    "url": "#",
  },
  Object {
    "anchor": undefined,
    "content": "    var robots = require('robots')      , parser = new robots.RobotsParser();     // for example:    //    // $ curl -s http://nodeguide.ru/robots.txt    //    // User-agent: Google-bot    // Disallow: /     // Crawl-delay: 2    //    // User-agent: *    // Disallow: /    // Crawl-delay: 2     parser.setUrl('http://nodeguide.ru/robots.txt', function(parser, success) {      if(success) {        var GoogleBotDelay = parser.getCrawlDelay(\\"Google-bot\\");        // ...      }    });",
    "level": undefined,
    "parents": Array [
      "robots.js",
      "Usage",
    ],
    "tags": Array [],
    "type": "text",
    "url": "#",
  },
  Object {
    "anchor": undefined,
    "content": "An example of passing options to the HTTP request:",
    "level": undefined,
    "parents": Array [
      "robots.js",
      "Usage",
    ],
    "tags": Array [],
    "type": "text",
    "url": "#",
  },
  Object {
    "anchor": undefined,
    "content": "var options = {  headers:{    Authorization:\\"Basic \\" + new Buffer(\\"username:password\\").toString(\\"base64\\")}} var robots = require('robots')  , parser = new robots.RobotsParser(null, options); parser.setUrl('http://nodeguide.ru/robots.txt', function(parser, success) {  ...});",
    "level": undefined,
    "parents": Array [
      "robots.js",
      "Usage",
    ],
    "tags": Array [],
    "type": "text",
    "url": "#",
  },
  Object {
    "anchor": null,
    "content": "API",
    "level": 2,
    "parents": Array [
      "robots.js",
    ],
    "tags": Array [],
    "type": "lvl",
    "url": "#",
  },
  Object {
    "anchor": undefined,
    "content": "RobotsParser — main class. This class provides a set of methods to read,
    parse and answer questions about a single robots.txt file.",
    "level": undefined,
    "parents": Array [
      "robots.js",
      "API",
    ],
    "tags": Array [],
    "type": "text",
    "url": "#",
  },
  Object {
    "anchor": undefined,
    "content": "setUrl(url, read) — sets the URL referring to a robots.txt file.
      by default, invokes read() method.
      If read is a function, it is called once the remote file is downloaded and parsed, and it
      takes in two arguments: the first is the parser itself, and the second is a boolean
      which is True if the the remote file was successfully parsed.",
    "level": undefined,
    "parents": Array [
      "robots.js",
      "API",
    ],
    "tags": Array [],
    "type": "text",
    "url": "#",
  },
  Object {
    "anchor": undefined,
    "content": "read(after_parse) — reads the robots.txt URL and feeds it to the parser",
    "level": undefined,
    "parents": Array [
      "robots.js",
      "API",
    ],
    "tags": Array [],
    "type": "text",
    "url": "#",
  },
  Object {
    "anchor": undefined,
    "content": "parse(lines) — parse the input lines from a robots.txt file",
    "level": undefined,
    "parents": Array [
      "robots.js",
      "API",
    ],
    "tags": Array [],
    "type": "text",
    "url": "#",
  },
  Object {
    "anchor": undefined,
    "content": "canFetch(userAgent, url, callback) — using the parsed robots.txt decide if
      userAgent can fetch url. Callback function:
      function callback(access, url, reason) { ... }
      where:
      
        access — can this url be fetched. true/false.
        url — target url
        reason — reason for access. Object:
          
            type — valid values: 'statusCode', 'entry', 'defaultEntry', 'noRule'
            entry — an instance of lib/Entry.js:. Only for types: 'entry', 'defaultEntry'
            statusCode — http response status code for url. Only for type 'statusCode'
          
        
      
    ",
    "level": undefined,
    "parents": Array [
      "robots.js",
      "API",
    ],
    "tags": Array [],
    "type": "text",
    "url": "#",
  },
  Object {
    "anchor": undefined,
    "content": "canFetchSync(userAgent, url) — using the parsed robots.txt decide if
      userAgent can fetch url. Return true/false.",
    "level": undefined,
    "parents": Array [
      "robots.js",
      "API",
    ],
    "tags": Array [],
    "type": "text",
    "url": "#",
  },
  Object {
    "anchor": undefined,
    "content": "getCrawlDelay(userAgent) — returns Crawl-delay for the certain userAgent",
    "level": undefined,
    "parents": Array [
      "robots.js",
      "API",
    ],
    "tags": Array [],
    "type": "text",
    "url": "#",
  },
  Object {
    "anchor": undefined,
    "content": "getSitemaps(sitemaps) — gets Sitemaps from parsed robots.txt",
    "level": undefined,
    "parents": Array [
      "robots.js",
      "API",
    ],
    "tags": Array [],
    "type": "text",
    "url": "#",
  },
  Object {
    "anchor": undefined,
    "content": "getDisallowedPaths(userAgent) — gets paths explictly disallowed for the user agent specified AND *",
    "level": undefined,
    "parents": Array [
      "robots.js",
      "API",
    ],
    "tags": Array [],
    "type": "text",
    "url": "#",
  },
  Object {
    "anchor": null,
    "content": "License",
    "level": 2,
    "parents": Array [
      "robots.js",
    ],
    "tags": Array [],
    "type": "lvl",
    "url": "#",
  },
  Object {
    "anchor": undefined,
    "content": "See LICENSE
    file.",
    "level": undefined,
    "parents": Array [
      "robots.js",
      "License",
    ],
    "tags": Array [],
    "type": "text",
    "url": "#",
  },
  Object {
    "anchor": null,
    "content": "Resources",
    "level": 1,
    "parents": Array [
      "robots.js",
    ],
    "tags": Array [],
    "type": "lvl",
    "url": "#",
  },
  Object {
    "anchor": undefined,
    "content": "Robots.txt Specifications by Google",
    "level": undefined,
    "parents": Array [
      "robots.js",
      "Resources",
    ],
    "tags": Array [],
    "type": "text",
    "url": "#",
  },
  Object {
    "anchor": undefined,
    "content": "Robots.txt parser for python",
    "level": undefined,
    "parents": Array [
      "robots.js",
      "Resources",
    ],
    "tags": Array [],
    "type": "text",
    "url": "#",
  },
  Object {
    "anchor": undefined,
    "content": "A Standard for Robot Exclusion",
    "level": undefined,
    "parents": Array [
      "robots.js",
      "Resources",
    ],
    "tags": Array [],
    "type": "text",
    "url": "#",
  },
]
`;
